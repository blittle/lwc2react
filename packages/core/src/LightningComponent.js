import React from 'react';
import ObservableMembrane from 'observable-membrane';

export default class LightningComponent extends React.Component {
  constructor(template) {
    super();

    const membrane = new ObservableMembrane({
      valueMutated: () => {
        if (this.mounted) this.forceUpdate();
      },
    });

    if (!this.__s) {
      this.__s = membrane.getProxy(Object.assign({}, this.props));
    }

    this.template = React.createRef();

    template.customEvents.forEach((event) => {
      const ref = event[2];
      this[ref] = React.createRef();
    });
  }

  componentDidUpdate(prevProps) {
    this.updateMembrane(prevProps);
  }

  componentDidMount(template) {
    this.mounted = true;
    this.updateMembrane();
    this.addLocalStyles(template);
    this.addEventHandlers(template);
  }

  componentWillUnmount(template) {
    this.mounted = false;
    this.cleanupLocalStyles(template);
    this.cleanupEventHandlers(template);
  }

  addLocalStyles(template) {
    this.stylesheets = [];
    template.stylesheets.forEach((stylesheet) => {
      const sheet = document.createElement('style');
      sheet.type = 'text/css';
      sheet.textContent = stylesheet(
        '[' + template.stylesheetTokens.hostAttribute.toLowerCase() + ']',
        '[' + template.stylesheetTokens.shadowAttribute.toLowerCase() + ']',
        null
      );
      document.head.appendChild(sheet);
      this.stylesheets.push(sheet);
    });
  }

  addEventHandlers(template) {
    template.customEvents.forEach((event) => {
      const name = event[0];
      const cb = event[1];
      const ref = event[2];
      this[cb] = this[cb].bind(this);
      if (ref !== 'template') {
        this[ref].current.template.current.addEventListener(name, this[cb]);
      } else {
        this[ref].current.addEventListener(name, this[cb]);
      }
    });
  }

  cleanupLocalStyles() {
    this.stylesheets.forEach((sheet) => {
      if (sheet.parentNode) sheet.parentNode.removeChild(sheet);
    });
  }

  cleanupEventHandlers(template) {
    template.customEvents.forEach((event) => {
      const name = event[0];
      const cb = event[1];
      const ref = event[3];

      if (ref !== 'template') {
        this[ref].current.template.current.removeEventListener(name, this[cb]);
      } else {
        this[ref].current.removeEventListener(name, this[cb]);
      }
    });
  }

  updateMembrane(prevProps = {}) {
    Object.keys(this.props).forEach((prop) => {
      if (prevProps[prop] !== this.props[prop])
        this.__s[prop] = this.props[prop];
    });
  }
}
