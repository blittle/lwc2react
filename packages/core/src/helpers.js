import { parse, types } from 'recast';

// const b = types.builders;
const n = types.namedTypes;

export function removeLWCCode(ast) {
  const lwcIndex1 = ast.program.body.findIndex(
    (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
  );
  ast.program.body[lwcIndex1] = parse(
    'import React from "react"'
  ).program.body[0];

  let lwcIndex2 = ast.program.body.findIndex(
    (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
  );

  while (lwcIndex2 !== -1) {
    ast.program.body.splice(lwcIndex2, 1);
    lwcIndex2 = ast.program.body.findIndex(
      (el) => n.ImportDeclaration.check(el) && el.source.value === 'lwc'
    );
  }

  const registerDecoratorsIndex = ast.program.body.findIndex(
    (el) =>
      n.ExpressionStatement.check(el) &&
      n.CallExpression.check(el.expression) &&
      el.expression.callee.name === '_registerDecorators'
  );

  if (registerDecoratorsIndex !== -1) {
    ast.program.body.splice(registerDecoratorsIndex, 1);
  }
}
