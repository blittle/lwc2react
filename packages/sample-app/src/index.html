<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width" />
    <title></title>
    <link rel="stylesheet" href="/css/global.css" />
  </head>
  <body>
    <div id="app"></div>
    <script type="module" src="dist/index.js"></script>
  </body>
</html>
