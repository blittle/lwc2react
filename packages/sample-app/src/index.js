import MyApp from 'my/home';
import React from 'react';
import ReactDOM from 'react-dom';

ReactDOM.render(React.createElement(MyApp), document.getElementById('app'));
