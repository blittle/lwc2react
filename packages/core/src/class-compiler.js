import { parse, types, prettyPrint } from 'recast';
import { removeLWCCode } from './helpers';

const b = types.builders;
const n = types.namedTypes;

function getClassAST(ast) {
  const exportStatement = ast.program.body.find((el) =>
    n.ExportDefaultDeclaration.check(el)
  );
  const exportName = (exportStatement?.declaration?.arguments || [])[0]?.name;

  return (
    ast.program.body.find(
      (element) =>
        n.ClassDeclaration.check(element) &&
        element.superClass.name === 'LightningElement'
    ) ||
    ast.program.body.find(
      (element) =>
        n.ClassDeclaration.check(element) && element.id.name === exportName
    )
  );
}

function getPublicProps(ast) {
  const decoratorsCall = ast.program.body.find(
    (el) =>
      n.ExpressionStatement.check(el) &&
      n.CallExpression.check(el.expression) &&
      el.expression.callee.name === '_registerDecorators'
  );
  const publicProps = decoratorsCall?.expression?.arguments[1]?.properties?.find(
    (prop) => prop.key.name === 'publicProps'
  );
  return publicProps?.value?.properties?.map((prop) => prop.key.name) ?? [];
}

function getClassMethods(classAST) {
  return classAST.body.body
    .filter((el) => n.MethodDefinition.check(el) && el.kind === 'method')
    .map((method) => method.key.name);
}

function convertClass(classAST, publicProps, templateIdentifier) {
  const classMethods = getClassMethods(classAST);

  if (classAST.superClass.name === 'LightningElement') {
    classAST.superClass = b.identifier('React.Component');
  }

  convertMethods(classAST, publicProps, classMethods);
  buildRenderMethod(classAST, templateIdentifier);
  buildStyles(classAST, templateIdentifier);
  convertConstructorBlock(classAST, publicProps, classMethods);
  convertLifeCycleMethods(classAST);
}

function convertConstructorBlock(classAST, publicProps, classMethods) {
  const constructorAST = classAST.body.body.find(
    (prop) => prop.kind === 'constructor'
  );

  if (constructorAST) {
    constructorAST.value.body = processBlock(constructorAST.value.body, {
      inConstructor: true,
      publicProps,
      classMethods,
    });

    const setup = parse(`
      const membrane = new ObservableMembrane({
        valueMutated: () => {
          if (this.mounted) this.forceUpdate()
        }
      });

      if (!this.__s) {
        this.__s = membrane.getProxy({
          ${classMethods.map((method) => `${method}: this.${method}`)}
        });
      }
      this.template = React.createRef();
    `);

    // initialize empty state
    constructorAST.value.body.body.splice(1, 0, ...setup.program.body);
  }
}

function buildRenderMethod(classAST, templateIdentifier) {
  // maybe extending `this` like this is not okay?
  const render = parse(`function render() {
    return ${templateIdentifier}(Object.assign(this, this.props, this.__s))
  }`);
  classAST.body.body.push(
    b.methodDefinition(
      'method',
      b.identifier('render'),
      b.functionExpression(
        b.identifier('render'),
        [],
        render.program.body[0].body
      )
    )
  );
}

function getMethodIndex(classAST, name) {
  return classAST.body.body.findIndex(
    (el) => n.MethodDefinition.check(el) && el.key.name === name
  );
}

function convertLifeCycleMethods(classAST) {
  const componentDidMount = classAST.body.body.find(
    (el) => n.MethodDefinition.check(el) && el.key.name === 'componentDidMount'
  );
  const componentWillUnmount = classAST.body.body.find(
    (el) =>
      n.MethodDefinition.check(el) && el.key.name === 'componentWillUnmount'
  );
  const connectedCallbackIndex = getMethodIndex(classAST, 'connectedCallback');
  let disconnectedCallbackIndex = getMethodIndex(
    classAST,
    'disconnectedCallback'
  );

  const renderedCallback = classAST.body.body.find(
    (el) => n.MethodDefinition.check(el) && el.key.name === 'renderedCallback'
  );
  const errorCallback = classAST.body.body.find(
    (el) => n.MethodDefinition.check(el) && el.key.name === 'errorCallback'
  );

  if (connectedCallbackIndex !== -1) {
    const connectedCallback = classAST.body.body[connectedCallbackIndex];
    componentDidMount.value.body.body.push(
      ...connectedCallback.value.body.body
    );
    classAST.body.body.splice(connectedCallbackIndex, 1);
  }

  if (disconnectedCallbackIndex !== -1) {
    disconnectedCallbackIndex = getMethodIndex(
      classAST,
      'disconnectedCallback'
    );
    const disconnectedCallback = classAST.body.body[disconnectedCallbackIndex];
    componentWillUnmount.value.body.body.push(
      ...disconnectedCallback.value.body.body
    );
    classAST.body.body.splice(disconnectedCallbackIndex, 1);
  }

  if (renderedCallback) {
    renderedCallback.key = b.identifier('componentDidUpdate');
  }

  if (errorCallback) {
    errorCallback.key = b.identifier('componentDidCatch');
  }
}

function processBlock(block, options = { inConstructor: false }) {
  block.body = block.body
    .map((element) => processElement(element, options))
    .filter(Boolean);
  return block;
}

function processAssignmentExpression(assignment, options) {
  const expression = assignment.expression;

  expression.left = processExpression(expression.left, options);
  expression.right = processExpression(expression.right, options);

  return assignment;
}

function processUpdateExpression(element, options) {
  element.expression.argument = processExpression(
    element.expression.argument,
    options
  );
  return element;
}

function processReturnStatement(returnStatement, options) {
  returnStatement.argument = processExpression(
    returnStatement.argument,
    options
  );
  return returnStatement;
}

function getTopOfMemberExpression(memberExpression) {
  if (n.MemberExpression.check(memberExpression.object)) {
    return getTopOfMemberExpression(memberExpression.object);
  } else {
    return memberExpression;
  }
}

function processCallExpression(callExpression, options) {
  if (n.MemberExpression.check(callExpression.expression.callee)) {
    const member = getTopOfMemberExpression(callExpression.expression.callee);

    if (
      n.ThisExpression.check(member.object) &&
      !options.classMethods.includes(member.property.name)
    ) {
      if (member.property.name === 'template') {
        member.property.name = 'template.current';
      } else if (
        member.property.name === 'dispatchEvent' ||
        member.property.name === 'addEventListener' ||
        member.property.name === 'removeEventListener'
      ) {
        callExpression.expression.callee = b.identifier(
          'this.template.current.' + member.property.name
        );
      } else {
        callExpression.expression.callee = processExpression(
          callExpression.expression.callee,
          options
        );
      }
    }

    const lifeCycleMap = {
      renderedCallback: 'componentDidUpdate',
      connectedCallback: 'componentDidMount',
      errorCallback: 'componentDidCatch',
      disconnectedCallback: 'componentWillUnmount',
    };

    if (lifeCycleMap.hasOwnProperty(member.property.name)) {
      member.property.name = lifeCycleMap[member.property.name];
    }
  }

  callExpression.expression.arguments = callExpression.expression.arguments.map(
    (arg) => processExpression(arg, options)
  );

  return callExpression;
}

function processVariableDeclaration(element, options) {
  for (let i = 0; i < element.declarations.length; i++) {
    const declaration = element.declarations[i];

    element.declarations[i] = processExpression(
      element.declarations[i],
      options
    );

    if (n.VariableDeclarator.check(declaration)) {
      element.declarations[i] = b.variableDeclarator(
        declaration.id,
        processExpression(declaration.init, options)
      );
    }
  }
  return element;
}

function processExpression(expression, options) {
  if (n.MemberExpression.check(expression)) {
    if (n.ThisExpression.check(expression.object)) {
      return b.identifier(`this.__s.${expression.property.name}`);
    } else if (n.MemberExpression.check(expression.object)) {
      expression.object = processExpression(expression.object, options);
      return expression;
    }
  }

  if (n.CallExpression.check(expression)) {
    return processCallExpression({ expression }, options).expression;
  }

  if (n.UpdateExpression.check(expression)) {
    expression.argument = processExpression(expression.argument, options);
    return expression;
  }

  if (n.BinaryExpression.check(expression)) {
    expression.left = processExpression(expression.left, options);
    expression.right = processExpression(expression.right, options);
    return expression;
  } else if (n.ConditionalExpression.check(expression)) {
    expression.test = processExpression(expression.test, options);
    expression.consequent = processExpression(expression.consequent, options);
    expression.alternate = processExpression(expression.alternate, options);
  }

  return expression;
}

function processForLoop(element, options) {
  element.test = processExpression(element.test, options);
  element.init = processVariableDeclaration(element.init, options);
  element.update = processExpression(element.update, options);
  element.body = n.BlockStatement.check(element.body)
    ? processBlock(element.body, options)
    : processExpression(element.body, options);

  return element;
}

function processForOfLoop(element, options) {
  element.left = processVariableDeclaration(element.left, options);
  element.right = processExpression(element.right, options);
  element.body = n.BlockStatement.check(element.body)
    ? processBlock(element.body, options)
    : processExpression(element.body, options);

  return element;
}

function processIfStatement(element, options) {
  element.test = processExpression(element.test, options);

  element.consequent = n.BlockStatement.check(element.consequent)
    ? processBlock(element.consequent, options)
    : processExpression(element.consequent, options);

  if (element.alternate) {
    element.alternate = n.BlockStatement.check(element.alternate)
      ? processBlock(element.alternate, options)
      : processExpression(element.alternate, options);
  }

  return element;
}

function processElement(element, options) {
  if (n.ExpressionStatement.check(element)) {
    if (n.AssignmentExpression.check(element.expression)) {
      return processAssignmentExpression(element, options);
    }
    if (n.UpdateExpression.check(element.expression)) {
      return processUpdateExpression(element, options);
    }

    if (n.CallExpression.check(element.expression)) {
      return processCallExpression(element, options);
    }
  } else if (n.VariableDeclaration.check(element)) {
    return processVariableDeclaration(element, options);
  } else if (n.IfStatement.check(element)) {
    return processIfStatement(element, options);
  } else if (n.ReturnStatement.check(element)) {
    return processReturnStatement(element, options);
  } else if (n.ForStatement.check(element)) {
    return processForLoop(element, options);
  } else if (n.ForOfStatement.check(element)) {
    return processForOfLoop(element, options);
  }

  return element;
}

function buildStyles(classAST, templateIdentifier) {
  const didMountText = parse(`
    this.mounted = true;
    this.stylesheets = [];
    ${templateIdentifier}.stylesheets.forEach(stylesheet => {
        const sheet = document.createElement("style");
        sheet.type = "text/css";
        sheet.textContent = stylesheet(
          "[" + ${templateIdentifier}.stylesheetTokens.hostAttribute.toLowerCase() + "]",
          "[" + ${templateIdentifier}.stylesheetTokens.shadowAttribute.toLowerCase() + "]",
           null
        );
        document.head.appendChild(sheet);
        this.stylesheets.push(sheet);
    })

    ${templateIdentifier}.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this[ref] = this[ref].bind(this);
      this.template.current.addEventListener(name, this[ref]);
    });
  `);

  const willUnmountText = parse(`
    this.mounted = false;
    this.stylesheets.forEach(sheet => {
      if (sheet.parentNode) sheet.parentNode.removeChild(sheet);
    })

    ${templateIdentifier}.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this.template.current.removeEventListener(name, this[ref]);
    });
  `);

  classAST.body.body.push(
    b.methodDefinition(
      'method',
      b.identifier('componentDidMount'),
      b.functionExpression(
        b.identifier('componentDidMount'),
        [],
        b.blockStatement(didMountText.program.body)
      )
    )
  );

  classAST.body.body.push(
    b.methodDefinition(
      'method',
      b.identifier('componentWillUnmount'),
      b.functionExpression(
        b.identifier('componentWillUnmount'),
        [],
        b.blockStatement(willUnmountText.program.body)
      )
    )
  );
}

function convertMethods(classAST, publicProps, classMethods) {
  return classAST.body.body
    .filter(
      (property) =>
        property?.kind === 'method' ||
        property?.kind === 'get' ||
        property?.kind === 'set'
    )
    .map((method) => {
      return b.methodDefinition(
        method.kind,
        b.identifier(method.key.name),
        convertMethod(method.value, {
          inConstructor: false,
          publicProps,
          classMethods,
        })
      );
    });
}

function convertMethod(method, options) {
  method.body.body = method.body.body
    .map((element) => processElement(element, options))
    .filter(Boolean);

  return method;
}

export function compileClass(ast) {
  const classAST = getClassAST(ast);

  if (classAST) {
    const publicProps = getPublicProps(ast);

    removeLWCCode(ast);

    ast.program.body.unshift(
      parse("import ObservableMembrane from 'observable-membrane'").program
        .body[0]
    );

    const exportStatement = ast.program.body.find((el) =>
      n.ExportDefaultDeclaration.check(el)
    );
    const templateIdentifier = exportStatement.declaration.arguments[1].properties.find(
      (prop) => prop.key.name === 'tmpl'
    ).value.name;

    exportStatement.declaration = exportStatement.declaration.arguments[0];

    convertClass(classAST, publicProps, templateIdentifier);
  }

  return prettyPrint(ast, { tabWidth: 2 }).code;
}
