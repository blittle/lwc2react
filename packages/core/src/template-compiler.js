import { parse, types, prettyPrint } from 'recast';
import { removeLWCCode } from './helpers';
import { eventMap, attributeMap } from './dataMaps';

const b = types.builders;
const n = types.namedTypes;

function getTemplateAST(ast) {
  return ast.program.body.find(
    (element) =>
      n.FunctionDeclaration.check(element) && element.params[0].name === '$api'
  );
}

function buildReactCreateElements(
  templateAST,
  customEvents,
  templateIdentifier
) {
  const returnAst = templateAST.body.body.find((el) =>
    n.ReturnStatement.check(el)
  );

  return b.blockStatement([
    b.returnStatement(
      returnAst
        ? recurseElementTree(returnAst.argument.elements, templateIdentifier, {
            slots: false,
            topOfTree: true,
            customEvents,
          })
        : b.nullLiteral()
    ),
  ]);
}

function recurseElementTree(astArray, templateIdentifier, options) {
  const process = (element) => {
    if (n.Literal.check(element)) return element;

    if (n.ConditionalExpression.check(element)) {
      element.consequent = recurseElementTree(
        [element.consequent],
        templateIdentifier,
        {
          ...options,
          slots: false,
          topOfTree: false,
          ref: options.ref,
        }
      );
      element.alternate = recurseElementTree(
        [element.alternate],
        templateIdentifier,
        {
          ...options,
          slots: false,
          topOfTree: false,
          ref: options.ref,
        }
      );

      return element;
    }

    const callee = element?.callee?.name;
    if (callee === 'api_element') {
      return b.callExpression(b.identifier('React.createElement'), [
        element.arguments[0],
        buildProps(
          element,
          false,
          templateIdentifier,
          options.topOfTree,
          options.customEvents
        ),
        buildChildren(element, templateIdentifier, {
          ...options,
          slots: false,
          topOfTree: false,
          ref: options.ref,
        }),
      ]);
    } else if (callee === 'api_text') {
      return element.arguments[0]; // return the string literal
    } else if (callee === 'api_dynamic') {
      return element.arguments[0];
    } else if (callee === 'api_custom_element') {
      return b.callExpression(b.identifier('React.createElement'), [
        element.arguments[1],
        buildProps(
          element,
          true,
          templateIdentifier,
          options.topOfTree,
          options.customEvents
        ),
        buildChildren(element, templateIdentifier, {
          ...options,
          slots: true,
          topOfTree: false,
          ref: options.ref,
        }),
      ]);
    } else if (callee === 'api_slot') {
      return getChildrenFromSlot(
        element,
        templateIdentifier,
        options.customEvents
      );
    } else {
      throw new Error('cannot process: ' + callee);
    }
  };

  if (options.slots && astArray.length) {
    return b.objectExpression(
      astArray.map((slot) =>
        b.objectProperty(b.stringLiteral(getSlotName(slot)), process(slot))
      )
    );
  }

  if (astArray.length === 1) {
    return process(astArray[0]);
  } else {
    if (!astArray.length) {
      return b.nullLiteral();
    } else {
      return b.arrayExpression(astArray.map(process));
    }
  }
}

function getChildrenFromSlot(element, templateIdentifier, customEvents) {
  const childTernary = parse(
    `($cmp.props.children && $cmp.props.children["${element.arguments[0].value}"]) ? $cmp.props.children["${element.arguments[0].value}"] : null`
  ).program.body[0].expression;
  childTernary.alternate = recurseElementTree(
    element.arguments[2]?.elements,
    templateIdentifier,
    { slots: false, topOfTree: false, customEvents }
  );
  return childTernary;
}

function getSlotName(slot) {
  return (
    slot?.arguments[1]?.properties
      ?.find((prop) => prop.key.name === 'attrs')
      ?.value?.properties?.find((prop) => prop.key.value === 'slot')?.value
      ?.value ?? ''
  );
}

function buildChildren(element, templateIdentifier, options) {
  const child = element.arguments[element.arguments.length - 1];

  if (n.CallExpression.check(child)) {
    const callee = child.callee.name;

    if (callee === 'api_iterator') {
      const ref =
        child.arguments[0].object.name + '.' + child.arguments[0].property.name;
      const param = child.arguments[1].params[0].name;

      const mapIteration = parse(`${ref}.map(${param} => null)`).program.body[0]
        .expression;

      const returnValues = recurseElementTree(
        [child.arguments[1].body.body[0].argument],
        templateIdentifier,
        {
          ...options,
          slots: false,
          topOfTree: false,
          ref: param,
        }
      );

      mapIteration.arguments[0].body = returnValues;
      return mapIteration;
    } else {
      throw new Error('Cannot process: ' + callee);
    }
  }

  return recurseElementTree(child?.elements, templateIdentifier, options);
}

function buildProps(
  element,
  component,
  templateIdentifier,
  topOfTree,
  customEvents
) {
  const index = component ? 2 : 1;

  const classMap = element.arguments[index].properties.find(
    (prop) => prop.key.name === 'classMap'
  );
  const styleMap = element.arguments[index].properties.find(
    (prop) => prop.key.name === 'styleMap'
  );
  const attrMap = element.arguments[index].properties.find(
    (prop) => prop.key.name === 'attrs'
  );
  const propMap = element.arguments[index].properties.find(
    (prop) => prop.key.name === 'props'
  );
  const className = element.arguments[index].properties.find(
    (prop) => prop.key.name === 'className'
  );
  const events = element.arguments[index].properties.find(
    (prop) => prop.key.name === 'on'
  );
  const key = element.arguments[index].properties.find(
    (prop) => prop.key.name === 'key'
  );

  const props = [];

  props.push(
    b.objectProperty(
      b.identifier(
        `[${templateIdentifier}.stylesheetTokens.shadowAttribute.toLowerCase()]`
      ),
      b.stringLiteral('true')
    )
  );

  if (key) {
    props.push(buildKey(key));
  }

  if (topOfTree) {
    props.push(
      b.objectProperty(b.identifier('ref'), b.identifier('$cmp.template'))
    );
  }

  if (events) {
    events.value.properties.forEach((prop) => {
      const args = prop.value.right.right.arguments[0];
      if (args.object.name === '$cmp') {
        if (eventMap['on' + prop.key.value]) {
          // native event
          props.push(
            b.property(
              'init',
              b.identifier(eventMap['on' + prop.key.value] || prop.key.value),
              parse(`$cmp.${args.property.name}.bind($cmp)`).program.body[0]
                .expression
            )
          );
        } else {
          // custom event
          customEvents.push([prop.key.value, args.property.name]);
        }
      } else {
        throw new Error('Unable to handle bound event: ' + prop.key.value);
      }
    });
  }

  if (propMap) {
    propMap.value.properties.forEach((prop) => {
      props.push(processProp(prop));
    });
  }

  if (attrMap) {
    attrMap.value.properties.forEach((prop) => {
      props.push(processProp(prop));
    });
  }

  if (classMap) {
    props.push(
      b.objectProperty(
        b.identifier('className'),
        b.stringLiteral(getClassNameString(classMap.value))
      )
    );
  }

  if (styleMap) {
    props.push(b.objectProperty(b.identifier('style'), styleMap.value));
  }

  if (className) {
    props.push(b.objectProperty(b.identifier('className'), className.value));
  }

  return props.length ? b.objectExpression(props) : b.nullLiteral();
}

function processProp(prop) {
  prop.key = b.stringLiteral(attributeMap[prop.key.value] || prop.key.value);
  if (
    n.CallExpression.check(prop.value) &&
    prop.value.callee.name === 'api_scoped_id'
  ) {
    prop.value = prop.value.arguments[0];
  }
  return prop;
}

function buildKey(prop) {
  if (
    n.CallExpression.check(prop.value) &&
    prop.value.callee.name === 'api_key'
  ) {
    prop.value = prop.value.arguments[1];
  }
  return prop;
}

function getClassNameString(objectExpression) {
  return objectExpression.properties.map((prop) => prop.key.value).join(' ');
}

export function compileTemplate(ast) {
  const templateAST = getTemplateAST(ast);

  const exportStatement = ast.program.body.find((el) =>
    n.ExportDefaultDeclaration.check(el)
  );
  const templateIdentifier = exportStatement.declaration.arguments[0].name;

  const customEvents = [];

  removeLWCCode(ast);

  templateAST.params = [b.identifier('$cmp')];
  templateAST.body = buildReactCreateElements(
    templateAST,
    customEvents,
    templateIdentifier,
    null
  );

  exportStatement.declaration = exportStatement.declaration.arguments[0];

  ast.program.body.push(
    parse(templateIdentifier + '.customEvents=' + JSON.stringify(customEvents))
      .program.body[0]
  );

  return prettyPrint(ast, { tabWidth: 2 }).code;
}
