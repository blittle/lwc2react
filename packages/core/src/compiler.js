import { parse } from 'recast';
import { compileClass } from './class-compiler';
import { compileTemplate } from './template-compiler';

export function compile(id, source) {
  if (id.includes('@lwc/engine/dist/engine.js') || id.includes('wire-service'))
    return 'export default undefined';

  const ast = parse(source);

  if (id.endsWith('.html')) {
    return compileTemplate(ast);
  } else if (id.endsWith('.css')) {
    return source;
  } else {
    return compileClass(ast);
  }
}
