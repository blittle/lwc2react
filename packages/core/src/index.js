import pluginUtils from '@rollup/pluginutils';
import { compile } from './compiler';

export default function rollupLWC2ReactCompiler(pluginOptions = {}) {
  let { include, exclude, debug } = pluginOptions;

  const filter = pluginUtils.createFilter(include, exclude);

  return {
    name: 'rollup-lwc2react-compiler',
    async transform(src, id) {
      if (debug && filter(id)) {
        console.log('lwc2react -> ', id);
      }

      if (!filter(id)) {
        return;
      }

      let code;
      try {
        code = compile(id, src);
        if (code) {
          if (debug && id.includes(debug)) {
            console.log(src);
            console.log('===========');
            console.log(code);
          }
          return {
            code,
            map: null,
          };
        }
      } catch (error) {
        console.log('error processing lwc source: ', id);
        console.log(src);
        throw error;
      }
    },
  };
}
