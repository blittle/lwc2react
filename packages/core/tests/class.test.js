import { compile } from '../src/compiler';

describe('General Class Conversion Tests', function () {
  it('should convert a component that extends from something other than LightningElement', function () {
    const source = `
        import { registerDecorators as _registerDecorators } from "lwc";
        import _tmpl from "./button.html";
        import { registerComponent as _registerComponent } from "lwc";
        import { classSet } from 'lightning/utils';
        import { normalizeString as normalize } from 'lightning/utilsPrivate';
        import LightningPrimitiveButton from 'lightning/primitiveButton';
        import template from './button.html';
        /**
         * A clickable element used to perform an action.
         */
        class LightningButton extends LightningPrimitiveButton {
            constructor(...args) {
                super(...args);
            }
            connectedCallback() {
                this.test = 'hi';
            }
            render() {
                return template;
            }
        }

        export default _registerComponent(LightningButton, {
            tmpl: _tmpl
        });
      `;

    expect(compile('sample.js', source)).toMatchSnapshot();
  });

  it('should convert calls to this.template to this.template.current', function () {
    const source = `
        import { registerDecorators as _registerDecorators } from "lwc";
        import _tmpl from "./button.html";
        import { registerComponent as _registerComponent } from "lwc";
        import { classSet } from 'lightning/utils';
        import { normalizeString as normalize } from 'lightning/utilsPrivate';
        import LightningPrimitiveButton from 'lightning/primitiveButton';
        import template from './button.html';
        /**
         * A clickable element used to perform an action.
         */
        class LightningButton extends LightningPrimitiveButton {
            renderedCallback() {
                this.template.host.style.pointerEvents = this.disabled ? 'none' : '';
                this.template.querySelector('hi');
            }

            connectedCallback() {
                this.template.querySelector('hi');
            }

            get computedIconClass() {
                return classSet('slds-button__icon').add({
                    'slds-button__icon_left': this.normalizedIconPosition === 'left',
                    'slds-button__icon_right': this.normalizedIconPosition === 'right'
                }).toString();
            }
        }

        export default _registerComponent(LightningButton, {
            tmpl: _tmpl
        });
      `;

    expect(compile('sample.js', source)).toMatchSnapshot();
  });
});
