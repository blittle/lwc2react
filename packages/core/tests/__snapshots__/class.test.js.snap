// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`General Class Conversion Tests should convert a component that extends from something other than LightningElement 1`] = `
"import ObservableMembrane from \\"observable-membrane\\";
import React from \\"react\\";
import _tmpl from \\"./button.html\\";
import { classSet } from \\"lightning/utils\\";
import { normalizeString as normalize } from \\"lightning/utilsPrivate\\";
import LightningPrimitiveButton from \\"lightning/primitiveButton\\";
import template from \\"./button.html\\";

/**
         * A clickable element used to perform an action.
         */
class LightningButton extends LightningPrimitiveButton {
  constructor(...args) {
    super(...args);

    const membrane = new ObservableMembrane({
      valueMutated: () => {
        if (this.mounted)
          this.forceUpdate();
      }
    });

    if (!this.__s) {
      this.__s = membrane.getProxy({
        connectedCallback: this.connectedCallback,
        render: this.render
      });
    }

    this.template = React.createRef();
  }

  render() {
    return template;
  }

  render() {
    return _tmpl(Object.assign(this, this.props, this.__s));
  }

  componentDidMount() {
    this.mounted = true;
    this.stylesheets = [];

    _tmpl.stylesheets.forEach(stylesheet => {
      const sheet = document.createElement(\\"style\\");
      sheet.type = \\"text/css\\";

      sheet.textContent = stylesheet(
        \\"[\\" + _tmpl.stylesheetTokens.hostAttribute.toLowerCase() + \\"]\\",
        \\"[\\" + _tmpl.stylesheetTokens.shadowAttribute.toLowerCase() + \\"]\\",
        null
      );

      document.head.appendChild(sheet);
      this.stylesheets.push(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this[ref] = this[ref].bind(this);
      this.template.current.addEventListener(name, this[ref]);
    });

    this.__s.test = \\"hi\\";
  }

  componentWillUnmount() {
    this.mounted = false;

    this.stylesheets.forEach(sheet => {
      if (sheet.parentNode)
        sheet.parentNode.removeChild(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this.template.current.removeEventListener(name, this[ref]);
    });
  }
}

export default LightningButton;"
`;

exports[`General Class Conversion Tests should convert calls to this.template to this.template.current 1`] = `
"import ObservableMembrane from \\"observable-membrane\\";
import React from \\"react\\";
import _tmpl from \\"./button.html\\";
import { classSet } from \\"lightning/utils\\";
import { normalizeString as normalize } from \\"lightning/utilsPrivate\\";
import LightningPrimitiveButton from \\"lightning/primitiveButton\\";
import template from \\"./button.html\\";

/**
         * A clickable element used to perform an action.
         */
class LightningButton extends LightningPrimitiveButton {
  componentDidUpdate() {
    this.__s.template.host.style.pointerEvents = this.__s.disabled ? \\"none\\" : \\"\\";
    this.template.current.querySelector(\\"hi\\");
  }

  get computedIconClass() {
    return classSet(\\"slds-button__icon\\").add({
      \\"slds-button__icon_left\\": this.normalizedIconPosition === \\"left\\",
      \\"slds-button__icon_right\\": this.normalizedIconPosition === \\"right\\"
    }).toString();
  }

  render() {
    return _tmpl(Object.assign(this, this.props, this.__s));
  }

  componentDidMount() {
    this.mounted = true;
    this.stylesheets = [];

    _tmpl.stylesheets.forEach(stylesheet => {
      const sheet = document.createElement(\\"style\\");
      sheet.type = \\"text/css\\";

      sheet.textContent = stylesheet(
        \\"[\\" + _tmpl.stylesheetTokens.hostAttribute.toLowerCase() + \\"]\\",
        \\"[\\" + _tmpl.stylesheetTokens.shadowAttribute.toLowerCase() + \\"]\\",
        null
      );

      document.head.appendChild(sheet);
      this.stylesheets.push(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this[ref] = this[ref].bind(this);
      this.template.current.addEventListener(name, this[ref]);
    });

    this.template.current.querySelector(\\"hi\\");
  }

  componentWillUnmount() {
    this.mounted = false;

    this.stylesheets.forEach(sheet => {
      if (sheet.parentNode)
        sheet.parentNode.removeChild(sheet);
    });

    _tmpl.customEvents.forEach(event => {
      const name = event[0];
      const ref = event[1];
      this.template.current.removeEventListener(name, this[ref]);
    });
  }
}

export default LightningButton;"
`;
